src/main/java/bio/terra/stairway/FlightDao.java

```diff
@@ -682,32 +755,54 @@ private void fillFlightContexts(Connection connection, List<FlightContext> fligh
   }
 
   /**
-   * Return flight state for a single flight
+   * Return flight state for a single flight. The status of completed flight is limited by
+   * completeFlightAvailable property.
    *
    * @param flightId flight to get
    * @return FlightState for the flight
+   * @throws FlightNotFound - no flight
+   * @throws DatabaseOperationException - unexpected error in the operation
+   * @throws InterruptedException - interrupt
    */
   FlightState getFlightState(String flightId)
       throws DatabaseOperationException, InterruptedException {
-    return DbRetry.retry("flight.getFlightState", () -> getFlightStateInner(flightId));
+    return DbRetry.retry("flight.getFlightState", () -> getFlightStateInner(flightId, true));
+  }
+
+  FlightState controlGetFlightState(String flightId)
+      throws DatabaseOperationException, InterruptedException {
+    return DbRetry.retry("flight.getFlightState", () -> getFlightStateInner(flightId, false));
   }
 
-  private FlightState getFlightStateInner(String flightId)
+  private FlightState getFlightStateInner(String flightId, boolean enforceVisibilityLimit)
       throws SQLException, DatabaseOperationException {
 
-    final String sqlOneFlight =
+    final String sqlOneFlightBase =
         "SELECT stairway_id, flightid, submit_time, "
             + " completed_time, output_parameters, status, serialized_exception"
             + " FROM "
             + FLIGHT_TABLE
             + " WHERE flightid = :flightId";
 
+    boolean enforceVisibility = enforceVisibilityLimit && (flightVisibilitySeconds > 0);
+
+    String sqlOneFlight = sqlOneFlightBase;
+    if (enforceVisibility) {
+      sqlOneFlight =
+          sqlOneFlightBase
+              + " AND (completed_time IS NULL OR completed_time + CAST(:visibility AS INTERVAL) > CURRENT_TIMESTAMP)";
```

Comment:
nit: Maybe inline the ':visibility seconds' or move the whole interval cast to the oneFlightStatement.setString? As-is the split between here and string.format was a little surprising to me.